modules/dnn/src/cuda4dnn/primitives/depth_space_ops.hpp

```diff
@@ -0,0 +1,76 @@
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef OPENCV_DNN_SRC_CUDA4DNN_PRIMITIVES_DEPTH_SPACE_OPS_HPP
+#define OPENCV_DNN_SRC_CUDA4DNN_PRIMITIVES_DEPTH_SPACE_OPS_HPP
+
+#include "../../op_cuda.hpp"
+
+#include "../csl/stream.hpp"
+#include "../csl/tensor.hpp"
+#include "../csl/tensor_ops.hpp"
+#include "../csl/memory.hpp"
+#include "../kernels/permute.hpp"
+
+#include <utility>
+
+namespace cv { namespace dnn { namespace cuda4dnn {
+
+    template <class T>
+    class DepthSpaceOps final : public CUDABackendNode {
+    public:
+        using wrapper_type = GetCUDABackendWrapperType<T>;
+
+        DepthSpaceOps(csl::Stream stream_, const std::vector<int> &internal_shape_,
+                     const std::vector<size_t> &permutation_)
+            : stream(std::move(stream_)), internal_shape(internal_shape_),
+              permutation(permutation_)
+        {
+            transposed_internal_shape = std::vector<int>(internal_shape.size());
+            for (size_t i = 0; i < permutation.size(); i++) {
+                transposed_internal_shape[i] = internal_shape[permutation[i]];
+            }
+
+            size_t num_elements = std::accumulate(internal_shape.begin(), internal_shape.end(), 1, std::multiplies<size_t>());
+            csl::WorkspaceBuilder builder;
+            builder.require<T>(num_elements);
+            scratch_mem_in_bytes = builder.required_workspace_size();
+        }
+
+        void forward(const std::vector<cv::Ptr<BackendWrapper>> &inputs,
+                     const std::vector<cv::Ptr<BackendWrapper>> &outputs,
+                     csl::Workspace &workspace) override {
+            CV_CheckEQ(inputs.size(), size_t(1), "DepthSpaceOps: only one input is accepted");
+            CV_CheckEQ(outputs.size(), size_t(1), "DepthSpaceOps: only one output is accepted");
+
+            auto input_wrapper = inputs.front().dynamicCast<wrapper_type>();
+            auto input = input_wrapper->getView();
+            CV_CheckEQ(input.rank(), size_t(4), "DepthSpaceOps: input needs to be 4-dimensional [N, C, H, W]");
+            auto output_wrapper = outputs.front().dynamicCast<wrapper_type>();
+            auto output = output_wrapper->getSpan();
+            auto ws_allocator = csl::WorkspaceAllocator(workspace);
+            auto transposed_internal = ws_allocator.get_tensor_span<T>(transposed_internal_shape.begin(), transposed_internal_shape.end());
+
+            // Call reshape on input so that it has the correct shape for permutation
+            input.reshape(internal_shape.begin(), internal_shape.end());
+            kernels::permute(stream, transposed_internal, input, permutation);
+            // Only copying is needed as output already has the expected shape
+            auto t = csl::TensorView<T>(transposed_internal);
+            csl::memcpy(output.get(), t.get(), output.size(), stream);
```

Comment:
Could we get rid of `memcpy` here?

Reply:
Technically we can, but we will have to make a custom copy of `kernels::permute` because it does not work with tensors of different ranks.

https://github.com/opencv/opencv/blob/acd16e384fe6253fb082afe33539ede6506a98e1/modules/dnn/src/cuda/permute.cu#L136-L144